### PR TITLE
RFC: make stalebot less manual work

### PR DESCRIPTION
**Pull request by vtjnash**
With the current settings, I need to either opt out entirely from the benefits of stalebot with "no-stale", or handle an additional continual stream of email notifications to avoid them falling off my list. I propose we make the "close" operation a manual step, such that "stale" operates instead like an automation of the existing "stalled" label.

Since it seems that each application of "stale" need be reviewed anyways, it seems equally reasonable to expect the reviewer to click the Github close button instead of the no-stale button and deleting the comment. This also means folks can take more than a one-week (daysToClose: 7) vacation, and the queue (https://github.com/libuv/libuv/labels/stale) will still be sitting there waiting for whenever you return.

### CODE REVIEW

**Comment by saghul** [comment]
Why is this such a big deal to you? So far I've only been happy with it everywhere I've used it, but maybe it's just me.

**Comment by cjihrig** [comment]
Would an acceptable middle ground be to bump `daysToClose` to something larger?

I haven't weighed in on any of the stalebot stuff so far because I have mixed feelings about it in general. But, as someone who has periodically gone through old issues to close them, I have to say that it removes some awkwardness and feeling like a jerk (not to mention the work of rechecking the same issues over and over).

**Comment by vtjnash** [comment]
I think I’m in a position where this bot is maximally inconvenient to me, as a heavy user and non-nodejs maintainer I’m in charge of some dozen patches and other bug fixes. But I don’t want to make this personal. The original PR said this was enabled as an experiment, and tried to make this clearly as an RFC specifically to encourage healthy debate and discussion over possible improvements. I have seen some good results from this bot. In other ways, it feels like it hasn’t lived up to its premise. But I've also found my opinions changed since watching the results of #2410.

For me, as a user and side contributor to libuv, this bot has been fairly negative experience. I’ll try to be concise at explaining why, even though I think some of the reasons are a bit complex to explain so I know this will get long. First let me answer cjihrig’s questions directly:

> Would an acceptable middle ground be to bump daysToClose to something larger?

Yes, it would be better. 7 days is really bad. To put this into perspective, I’ve historically seemed to have had about 1-2 day/month to work on libuv. With this set to 90/7, that means most of my time might now end up going into just reading the bot notifications. And each one suddenly has an artificial 1 week deadline.

I started this PR with the intent to just increase that time. But as I did, I took a step back too and thought “what is this bot really trying to accomplish?” And then ended up with this.

> it removes some awkwardness and feeling like a jerk

I don’t think it’s better having a bot do it, but just shifts the pain from the maintainer hitting close to the user who submitted the issue or PR. By analogy, do you feel better knowing it’s a robot that answered your telephone call or denied your credit application? Yes, we know that it’s better for the company.

> the work of rechecking the same issues over and over

Yes, this bot is making that worse, not better for me. Now instead of wondering what PRs were forgotten and hoping the authors will bump the important ones, I get a repeat email about them every few months. Or I can disable the bot by apply the “not-stale” label, and be right back where we started before enabling the bot.

I think it’d help to list out the goals for the bot. Off the top of my head, this is my list of things that I think it does or that I want it to do or that I assume the bot exists to do:

1. avoid forgetting about merging PRs
2. avoid forgetting fixing critical issues
3. close invalid issues (e.g. not enough information provided)
4. helps users landing here to submit a bug or feature find out if it’s already known but needs help
5. hit “inbox zero”

Of these, I think the current bot does 3 well, 1 & 2 & 5 mediocre, and 4 badly. Any more to add?

Without the bot, I’d say we do 1-4 mediocre and 5 badly. I’d also say that 5 is good for some contexts, like personal email, but the wrong criteria for most software projects (and it should be the outcome, not the goal). With the availability of the “not-stale” tag, I think this is also the equilibrium we’ll also necessarily settle back into as all valid issues get marked.

I think 4 is a really important, and an underrated one. An open issue is much easier to find than a closed one. I think this contributes to several network effects:
  - surprise new contributors may pickup an existing issue and solve it
  - bug reports can accumulate to finally push a stale issue into an understood and solvable state
  - when closing one bug, it’s possible to search for other old issues that might also get fixed and add them to the tests
  - it’s more rewarding to make a PR that closes more issues, especially if they’ve been opened for a while

Therefore, I think we might achieve a better outcome with this PR. This should make it easy for someone to see that an issue is marked stale and click close (no comment required). Still perhaps not doing any of those well, but at least less mediocre than without the bot. In fact, I’d propose we also cut the daysUntilStale in half (45 days) so that the reminder happens much sooner. I’d also like to propose renaming the “not-stale” flag to  “needs help,” to make it more descriptive. Perhaps it might also use process-descriptive labels (needs-review, needs-edits, needs-decision, etc.)?

tl;dr; I think stalebot currently optimizes for the wrong goal, and this PR may help to better align its behavior with our goals for it and be less hostile to users.

**Comment by saghul** [comment]
Thanks for putting this together @vtjnash. I think the main disagreement (if we can call it that way) comes from having different motives / expectations.

I'm going to use the first person singular here, because I proposed the change, and because I don't know if those who approved / liked it share my resoning.

You mentioned this is *your* list of things the bot should be helping with. (emphasis on your)

> * avoid forgetting about merging PRs
> * avoid forgetting fixing critical issues
> * close invalid issues (e.g. not enough information provided)
> * helps users landing here to submit a bug or feature find out if it’s already known but needs help
> * hit “inbox zero”

It's a good list to begin with, but *I* (emphasis on I) don't share all those:

* I long ago declared "GH email bankruptcy" so hitting inbox zero is not a goal of mine
* Have we ever forgotten to fix critical issues?

More importantly (again, for me) the list lacks the one single reason I proposed stalebot in the first place: trying to help with "developer fatigue". I can't speak for everyone, but it does take a toll on me to work my ass off only to find the amount of issues hasn't decreased. It may sound silly, but I find the feeling of progress encouraging.

Disabling the closing of issues nullifies the usefulness of the bot (IMHO), so I'd rather keep it. I am, however not strongly opinionated on the timelines, so I'd be happy to change / double both if that helps others.

**Comment by vtjnash** [comment]
> I'm going to use the first person singular here

👍

> your list of things

My intent was to say this was _a_ list of things, which is a superset of things that I think _some_ bot could aid, and things that I assumed others want a bot to help do. And my hope was to have others add to the list.

> I long ago declared "GH email bankruptcy" so hitting inbox zero is not a goal of mine

I was thinking "inbox zero" and "amount of issues hasn't decreased" were the same when writing my list. I can see now how they aren't necessarily. Addressing developer fatigue is important too. It doesn't sound silly at all, as I used to also have the goal of watching the issue count decrease. Others have since forced me to revise my goal, to avoid going crazy: it simply didn't make sense for a large project with a growing user base. For myself, I now keep a record of the number and rate at which issues get solved, instead of looking at the ratio of issues getting closed. (for better or worse, for related reasons, it's been a private list which few people have access to)

> Have we ever forgotten to fix critical issues?

To be blunt? I would say yes. (And to be clear, I think this would be literally impossible, since "critical issue" isn't a well-defined set. Please don't take it personal, as I don't intend this as an attack in any way.) From the sidelines, it looks like libuv is pretty good and rapid at merging and fixing any regressions to Nodejs, but necessarily can't do as well at responding to other projects. For that reason, I'm currently maintaining about a dozen patches on top of libuv master (https://github.com/libuv/libuv/compare/master...JuliaLang:julia-uv2-1.29.1) to fix various issues or add essential features. Additionally I have a small number of PRs (https://github.com/libuv/libuv/pulls/vtjnash) which are stalled for various reasons.

> I'd be happy to change / double both if that helps others.

Sure, my alternate proposal would be to switch them and make daysUntilStale as low as 21 and daysUntilClose at least 180. This gives new PRs (and issues) a quicker reminder, but gives a larger window before they get discarded.

Somewhat separately, we could also consider adding additional tags than "not-stale" to the list, which give the high-level reason it was not making progress (needs-review, needs-edits, needs-formatting, needs-decision, good-first-issue, minor-improvement, etc.). This would hopefully aid someone sweeping through stale work to help them more easily solve (rather than just close) the issues.

**Comment by saghul** [comment]
> My intent was to say this was _a_ list of things, which is a superset of things that I think _some_ bot could aid, and things that I assumed others want a bot to help do. And my hope was to have others add to the list.

👍



> Others have since forced me to revise my goal, to avoid going crazy: it simply didn't make sense for a large project with a growing user base. For myself, I now keep a record of the number and rate at which issues get solved, instead of looking at the ratio of issues getting closed. (for better or worse, for related reasons, it's been a private list which few people have access to)

I'll need to improve on this, alas, it seems to last a while, at least for me. Thanks for sharing!

> Please don't take it personal, as I don't intend this as an attack in any way.)

❤️  Didn't, I appreciate the candid responses here!

> From the sidelines, it looks like libuv is pretty good and rapid at merging and fixing any regressions to Nodejs, but necessarily can't do as well at responding to other projects.

Yeah, that is true. At the same time, the cross-breeding of developers is a lot higher to Node so I guess it's somewhat proportional... Do you think there is anything specific (no matter how small) we could do to help Julia, for example?

> For that reason, I'm currently maintaining about a dozen patches on top of libuv master

Right. I feel you. I hope v2 comes even though it won't be in the form I thought it would. As time passed I came to realize that some of the changes proposed for v2 no longer even make sense, they just add churn for little gain (ie renaming uv_idle_t). We have many dependent projects, and "yeah use v2, it breaks everything" is not nice. We are now trapped in v1.x for a while but IIRC someone proposed to change the major cadence once we hit v2. I hope that can help reduce your maintenance burden, even if it won't happen in the short term.

> Sure, my alternate proposal would be to switch them and make daysUntilStale as low as 21 and daysUntilClose at least 180. This gives new PRs (and issues) a quicker reminder, but gives a larger window before they get discarded.

SGTM!

> Somewhat separately, we could also consider adding additional tags than "not-stale" to the list, which give the high-level reason it was not making progress (needs-review, needs-edits, needs-formatting, needs-decision, good-first-issue, minor-improvement, etc.). This would hopefully aid someone sweeping through stale work to help them more easily solve (rather than just close) the issues.

This makes total sense to me!

**Comment by bnoordhuis** [comment]
> We are now trapped in v1.x for a while but IIRC someone proposed to change the major cadence once we hit v2.

At the risk of derailing... I kind of like the thought of v1.x forever. Stable APIs are nice! With some thought and care, I think it could be made to work in a way that isn't too onerous/restrictive.

**Comment by saghul** [comment]
> At the risk of derailing... I kind of like the thought of v1.x forever. Stable APIs are nice! With some thought and care, I think it could be made to work in a way that isn't too onerous/restrictive.

Happy to move this discussion elsewhere, but I'll start here.

I think v1.x is not sustainable forever. v2.x (even in its current shape) could be It. Off tghe top of my head, there are 2 major things v2 fixes which v1.x will never be able to do:

- we forgot to add reserved fields to uv_loop_t
- Windows file handles vs fds

**Comment by bnoordhuis** [comment]
Yeah, let's move it to #2506. :-)

**Comment by saghul** [comment]
@vtjnash wanna update this so we can land it?

**Comment by vtjnash** [comment]
OK, PTAL at the new proposal

**Comment by saghul** [comment]
LGTM!
